### PR TITLE
[LETS-45] Send request to fetch a log page from page server.

### PR DESCRIPTION
**Pull request by danilailie**
http://jira.cubrid.org/browse/LETS-45

A new type of request message was added (SEND_LOG_PAGE_FETCH) that's sent from the Transaction Server to Page Server containing the page log id requested.

At the same time a debug system parameter was added, er_log_read_log_page, that when set, logs the log page requests on both side of the communication (Transaction and Page server)

Example of log snippets:
```
Transaction server:

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/transaction/log_page_buffer.c, line 1944
Sent request for log to Page Server. Page ID: 105 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/transaction/log_page_buffer.c, line 1944
Sent request for log to Page Server. Page ID: 106 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/transaction/log_page_buffer.c, line 1944
Sent request for log to Page Server. Page ID: 107 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/transaction/log_page_buffer.c, line 1944
Sent request for log to Page Server. Page ID: 108
```

```
Page server:

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/server/page_server.cpp, line 99
Received request for log from Transaction Server. Page ID: 105 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/server/page_server.cpp, line 99
Received request for log from Transaction Server. Page ID: 106 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/server/page_server.cpp, line 99
Received request for log from Transaction Server. Page ID: 107 

Time: 03/18/21 00:56:03.793 - DEBUG *** file /home/qa/tesseractdb/src/server/page_server.cpp, line 99
Received request for log from Transaction Server. Page ID: 108
```

### CODE REVIEW

File: src/broker/broker_config.c [hunk 1]
```diff
@@ -377,8 +377,7 @@ get_conf_string (int value, T_CONF_TABLE * conf_table)
  */
 static int
 broker_config_read_internal (const char *conf_file, T_BROKER_INFO * br_info, int *num_broker, int *br_shm_id,
-			     char *admin_log_file, char admin_flag, bool * acl_flag, char *acl_file,
-			     char *admin_err_msg)
+			     char *admin_log_file, char admin_flag, bool *acl_flag, char *acl_file, char *admin_err_msg)
```
Comment (ribram) [inline src/broker/broker_config.c:380]: Please make sure that indent versions is 2.2.11

File: src/server/page_server.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include "server_type.hpp"
 #include "system_parameter.h"
 
+#include <bitset>
```
Comment (ribram) [inline src/server/page_server.cpp:28]: <bitset> is not used.

File: src/server/page_server.cpp [hunk 3]
```diff
@@ -52,6 +53,10 @@ page_server::set_active_tran_server_connection (cubcomm::channel &&chn)
 					std::bind (&page_server::receive_log_prior_list, std::ref (*this),
 					    std::placeholders::_1));
 
+  m_ats_conn->register_request_handler (ats_to_ps_request::SEND_LOG_PAGE_FETCH,
+        std::bind(&page_server::receive_log_page_fetch, std::ref(*this),
```
Comment (ribram) [inline src/server/page_server.cpp:57]:
```suggestion
        std::bind (&page_server::receive_log_page_fetch, std::ref (*this),

```

File: src/server/page_server.cpp [hunk 4]
```diff
@@ -78,6 +83,22 @@ page_server::receive_log_prior_list (cubpacking::unpacker &upk)
...
+                LOG_PAGEID pageid;
+                std::string message;
+
+                upk.unpack_string(message);
+                memcpy(&pageid, message.c_str(), sizeof(pageid));
+                assert(message.size() == sizeof(pageid));
+
+                _er_log_debug (ARG_FILE_LINE, "Received request for log from Transaction Server. Page ID: %lld \n", pageid);
+        }
+}
```
Comment (ribram) [inline src/server/page_server.cpp:100]: Code is badly formatted. Please make sure that astyle is working.

File: src/server/page_server.cpp [hunk 5]
```diff
@@ -78,6 +83,22 @@ page_server::receive_log_prior_list (cubpacking::unpacker &upk)
...
+void
+page_server::receive_log_page_fetch (cubpacking::unpacker &upk)
+{
+        if (prm_get_bool_value (PRM_ID_ER_LOG_READ_LOG_PAGE))
+        {
+                LOG_PAGEID pageid;
+                std::string message;
+
+                upk.unpack_string(message);
+                memcpy(&pageid, message.c_str(), sizeof(pageid));
```
Comment (ribram) [inline src/server/page_server.cpp:95]: Use std:: namespace for standard functions.
```suggestion
                std::memcpy(&pageid, message.c_str(), sizeof(pageid));

```

File: src/transaction/log_page_buffer.c [hunk 6]
```diff
@@ -183,6 +185,8 @@ static int rv;
 
 /* LOG BUFFER STRUCTURE */
 
+constexpr size_t BIG_INT_SIZE = 8;
```
Comment (ribram) [inline src/transaction/log_page_buffer.c:188]: Move the BIG_INT_SIZE declaration in the scope that it is used.

File: src/transaction/log_page_buffer.c [hunk 7]
```diff
@@ -1927,6 +1931,23 @@ logpb_copy_page (THREAD_ENTRY * thread_p, LOG_PAGEID pageid, LOG_CS_ACCESS_MODE
     }
 
   /* Could not get from log page buffer cache */
+
+#if defined (SERVER_MODE)
+  /* Send a request to Page Server for the log. */
+  if (get_server_type () == SERVER_TYPE_TRANSACTION)
+    {
+      char buffer[BIG_INT_SIZE];
+      memcpy (buffer, &pageid, sizeof (pageid));
```
Comment (ribram) [inline src/transaction/log_page_buffer.c:1940]:
```suggestion
      std::memcpy (buffer, &pageid, sizeof (pageid));
```

**Review by ribram — COMMENTED** [review]
Please update the title to the format `[LETS-45] Send request to fetch a log page from page server`.

In the PR description you can use the formatting options that github offers. Code & text snippets can be bordered by three quotes.

File: src/transaction/log_page_buffer.c (hunk 7 above)
Comment (danilailie) [reply src/transaction/log_page_buffer.c:1940]: namespace "std" has no member "memcpy"C/C++(135)

Comment (ribram) [reply src/transaction/log_page_buffer.c:1940]: It is in the <cstring> header.

**Review by ribram — APPROVED** [review]

**Pull request merged**